Assets/Scripts/Weapon/HarpoonGun/HarpoonGun.cs

```diff
@@ -0,0 +1,94 @@
+using System.Collections.Generic;
+using Enemies;
+using UnityEngine;
+
+public class HarpoonGun : WeaponClass
+{
+
+    [Header("HarpoonGun Stats")]
+    [SerializeField] public int harpoons = 3;
+    [SerializeField] public float harpoonVelocity = 20;
+    [SerializeField] public float harpoonSpearGravityScale = 1;
+    [SerializeField] public GameObject harpoonSpearPrefab;
+
+    // Assuming harpoons isn't too great, otherwise will use HashSet
+    LinkedList<HarpoonSpear> spears = new LinkedList<HarpoonSpear>();
+
+    public override void Attack()
+    {
+        if (spears.Count >= harpoons) {
+            // Do nothing
+            return;
+        }
+
+        GameObject spearObject = Instantiate(harpoonSpearPrefab, transform.position, transform.rotation);
+        HarpoonSpear spear = spearObject.GetComponent<HarpoonSpear>();
+        spears.AddFirst(spear);
+        spear.Fire(this);
+    }
+
+    public override void UseSkill()
+    {
+        if (spears.Count == 0) {
+            return;
+        }
+
+        // Find the next most recent spear that has tagged enemy
+        Enemy tagged = null;
```

Comment:
spears.Where(spear => spear.TaggedEnemy != null)

Reply:
the LINQer